### PR TITLE
Service staking refactor

### PR DESCRIPTION
**Pull request by kupermind**
- Service staking implementation accounting for latest suggestions;
- Adding more tests.

### CODE REVIEW

File: contracts/staking/ServiceStakingBase.sol [hunk 1]
```diff
@@ -191,7 +281,6 @@ abstract contract ServiceStakingBase is IErrorsRegistries {
     /// @param eligibleServiceIds Service Ids eligible for rewards.
     /// @param eligibleServiceRewards Corresponding rewards for eligible service Ids.
     /// @param serviceIds All the staking service Ids.
-    /// @param serviceNonces Current service nonces.
```
Comment (mariapiamo) [inline contracts/staking/ServiceStakingBase.sol:194]: why this comment was removed?

File: contracts/staking/ServiceStakingBase.sol [hunk 2]
```diff
@@ -202,96 +291,115 @@ abstract contract ServiceStakingBase is IErrorsRegistries {
...
+
+                    // Get current service multisig nonce
+                    serviceNonces[i] = IMultisig(curInfo.multisig).nonce();
+
+                    // Calculate the liveness nonce ratio
+                    // Get the last service checkpoint: staking start time or the global checkpoint timestamp
+                    uint256 serviceCheckpoint = tsCheckpointLast;
+                    uint256 tsStart = curInfo.tsStart;
+                    // Adjust the service checkpoint time if the service was staking less than the current staking period
+                    if (tsStart > serviceCheckpoint) {
```
Comment (mariapiamo) [inline contracts/staking/ServiceStakingBase.sol:330]: With this approach, we are not forcing a service to be staked for the entire liveness period. As we did previously, service multisigs that pass the liveness check will be considered for incentives, even if they have been staked for less than the liveness period. Is that correct?

File: contracts/staking/ServiceStakingBase.sol [hunk 3]
```diff
@@ -202,96 +291,115 @@ abstract contract ServiceStakingBase is IErrorsRegistries {
...
+                        serviceCheckpoint = tsStart;
+                    }
+                    // Calculate the liveness ratio in 1e18 value
+                    uint256 ratio;
+                    // If the checkpoint was called in the exactly same block, the ratio is zero
```
Comment (mariapiamo) [inline contracts/staking/ServiceStakingBase.sol:335]: If the checkpoint was called in the exactly same block, we do not reach this part. No?

Comment (kupermind) [reply contracts/staking/ServiceStakingBase.sol:335]: I think it is possible if someone calls staking and checkpoint in the same block

File: contracts/staking/ServiceStakingBase.sol (hunk 1 above)
Comment (kupermind) [reply contracts/staking/ServiceStakingBase.sol:194]: It was from the previous PR when this was considered obsolete, I've changed that now.

File: contracts/staking/ServiceStakingBase.sol (hunk 2 above)
Comment (kupermind) [reply contracts/staking/ServiceStakingBase.sol:330]: That is correct, as the checkpoint is a centralization point and everybody eligible must account for their rewards.

File: contracts/staking/ServiceStakingBase.sol [hunk 4]
```diff
@@ -97,37 +146,56 @@ abstract contract ServiceStakingBase is IErrorsRegistries {
...
+        if (size > 0) {
+            for (uint256 i = 0; i < size; ++i) {
+                // Agent Ids must be unique and in ascending order
+                if (_stakingParams.agentIds[i] <= agentId) {
+                    revert WrongAgentId(_stakingParams.agentIds[i]);
+                }
+                agentId = _stakingParams.agentIds[i];
+                agentIds.push(agentId);
+            }
+        }
```
Comment (DavidMinarsch) [inline contracts/staking/ServiceStakingBase.sol:195]: Isn't this guaranteed by the registry already?

Comment (kupermind) [reply contracts/staking/ServiceStakingBase.sol:195]: Staking service agent Ids must match the staking contract configuration. We want to ensure at least the minimum order requirement.

**Review by 77ph — APPROVED** [review]

**Pull request merged**